src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs

```diff
@@ -22,57 +22,114 @@
+
+            // If the application retrieved the token from somewhere else, set it.
+            string token = accessTokenReceivedContext.Token;
+
+            // Bypass the default processing to pull the token from the Authorization header
+            //  if the event handled the request parsing.
+            if(!accessTokenReceivedContext.Handled)
+            {
+                string header = Request.Headers[HeaderNames.Authorization];
+                if (string.IsNullOrEmpty(header))
+                {
+                    return AuthenticateResult.Fail("Authentication failed because the bearer token " +
+                                                   "was missing from the 'Authorization' header.");
+                }
+
+                // Ensure that the authorization header contains the mandatory "Bearer" scheme.
+                // See https://tools.ietf.org/html/rfc6750#section-2.1
+                if (!header.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
+                {
+                    return AuthenticateResult.Fail("Authentication failed because an invalid scheme " +
+                                                   "was used in the 'Authorization' header.");
+                }
 
-            // Ensure that the authorization header contains the mandatory "Bearer" scheme.
-            // See https://tools.ietf.org/html/rfc6750#section-2.1
-            if (!header.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase)) {
-                return AuthenticateResult.Fail("Authentication failed because an invalid scheme " +
-                                               "was used in the 'Authorization' header.");
+                token = header.Substring("Bearer ".Length);
             }
 
-            var token = header.Substring("Bearer ".Length);
-            if (string.IsNullOrWhiteSpace(token)) {
+            if (string.IsNullOrWhiteSpace(token))
+            {
                 return AuthenticateResult.Fail("Authentication failed because the bearer token " +
-                                               "was missing from the 'Authorization' header.");
+                                                "was missing from the request. The default location" +
+                                                "is in the 'Authorization' header.");
             }
 
             // Try to resolve the authentication ticket from the distributed cache. If none
             // can be found, a new introspection request is sent to the authorization server.
             var ticket = await RetrieveTicketAsync(token);
-            if (ticket == null) {
+            if (ticket == null)
+            {
+                // Allow interception of the introspection retrieval process via events
+                var requestTokenIntrospectionContext = new RequestTokenIntrospectionContext(Context, Options, token);
+                await Options.Events.RequestTokenIntrospection(requestTokenIntrospectionContext);
```

Comment:
I wonder if this event offers enough flexibility. I think one should at least be able to alter the HTTP request message to add custom parameters before it sent by the introspection middleware.